src/mptool.f90

```diff
@@ -143,5 +290,343 @@ pure function cross_product(a,b)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_1darray
+
+
+  subroutine pread_2darray(filename,data2read)
+
+    ! arguments
+    character(len=*),intent(in) :: filename
+    real(mytype),intent(inout),allocatable :: data2read(:,:)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    local_size=size(data2read,2)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_2darray
+
+  ! this function return the number add from all processors before it
+  integer function prelay(number)
+
+    ! arguments
+    integer,intent(in) :: number
+
+    ! local data
+    integer :: table(nproc)
+    integer :: ierr,i
+
+    call mpi_allgather(number,1,mpi_integer,                              &
+                       table,1,mpi_integer,mpi_comm_world,ierr)
+
+    prelay=0
+    do i=1,nrank
+      prelay=prelay+table(i)
+    enddo
+    
+    return
+
+  end function prelay
   !
+  function linintp(xx1,xx2,yy1,yy2,xx) result(yy)
+    !
+    real(mytype),intent(in) :: xx1,xx2,xx
+    real(mytype),intent(in) ::  yy1(:,:,:),yy2(:,:,:)
+    real(mytype) :: yy(1:size(yy1,1),1:size(yy1,2),1:size(yy1,3))
+    !
+    real(mytype) :: var1
+    !
+    if(abs(xx-xx1)<1.e-16_mytype) then
+      yy=yy1
+    elseif(abs(xx-xx2)<1.e-16_mytype) then
+      yy=yy2
+    else
+      var1=(xx-xx1)/(xx2-xx1)
```

Comment:
I think the variable `var1` is not really needed

Reply:
Please add a statement when you resolve a conversation without any modification in the code

Reply:
use epsilon(1._mytype) instead

Reply:
I guess this has been resolved as well